Source/WebCore/page/NavigatorLoginStatus.cpp

```diff
@@ -47,40 +54,92 @@ ASCIILiteral NavigatorLoginStatus::supplementName()
 }
 
 void NavigatorLoginStatus::isLoggedIn(Navigator& navigator, Ref<DeferredPromise>&& promise)
 {
     NavigatorLoginStatus::from(navigator)->isLoggedIn(WTFMove(promise));
 }
 
-void NavigatorLoginStatus::setLoggedIn(Ref<DeferredPromise>&& promise)
+bool NavigatorLoginStatus::hasSameOrigin() const
 {
-    if (m_navigator.cookieEnabled())
-        promise->resolve();
-    else
-        promise->reject();
+    RefPtr document = m_navigator.document();
+    Ref origin = document->securityOrigin();
+    bool isSameSite = true;
+    for (RefPtr parentDocument = document->parentDocument(); parentDocument; parentDocument = parentDocument->parentDocument()) {
+        if (!origin->isSameOriginAs(parentDocument->securityOrigin())) {
+            isSameSite = false;
+            break;
+        }
+    }
+    return isSameSite;
 }
 
-void NavigatorLoginStatus::setLoggedOut(Ref<DeferredPromise>&& promise)
+void NavigatorLoginStatus::setLoginStatus(const String& status, Ref<DeferredPromise>&& promise)
 {
-    promise->resolve();
+    IsLoggedIn loginStatus;
+    if (status == "logged-in"_s)
+        loginStatus = IsLoggedIn::Yes;
+    else if (status == "logged-out"_s)
+        loginStatus = IsLoggedIn::No;
+    else {
+        promise->reject();
+        return;
+    }
+
+    RefPtr document = m_navigator.document();
+    if (!document && hasSameOrigin()) {
+        ASSERT_NOT_REACHED();
+        promise->reject();
+        return;
+    }
+    RefPtr frame = document->frame();
+    if (!frame || !frame->isMainFrame()) {
+        ASSERT_NOT_REACHED();
+        promise->reject();
+        return;
+    }
+    RefPtr page = frame->page();
+    if (!page) {
+        ASSERT_NOT_REACHED();
+        promise->reject();
+        return;
+    }
+    page->chrome().client().setLoginStatus(RegistrableDomain::uncheckedCreateFromHost(document->securityOrigin().host()), loginStatus, [weakThis = WeakPtr { *this }, promise = WTFMove(promise)] {
+        if (!weakThis)
+            return;
+        promise->resolve();
+    });
 }
 
 void NavigatorLoginStatus::isLoggedIn(Ref<DeferredPromise>&& promise)
 {
-    if (m_navigator.cookieEnabled())
-        promise->resolve<IDLBoolean>(true);
-    else
-        promise->resolve<IDLBoolean>(false);
+    RefPtr document = m_navigator.document();
+    if (!document) {
+        ASSERT_NOT_REACHED();
+        promise->reject();
+        return;
+    }
+    RefPtr frame = document->frame();
+    if (!frame || !frame->isMainFrame()) {
+        ASSERT_NOT_REACHED();
+        promise->reject();
+        return;
+    }
+    RefPtr page = frame->page();
+    if (!page) {
+        ASSERT_NOT_REACHED();
+        promise->reject();
+        return;
+    }
+    page->chrome().client().isLoggedIn(RegistrableDomain::uncheckedCreateFromHost(document->securityOrigin().host()), [weakThis = WeakPtr { *this }, promise = WTFMove(promise)] (bool isLoggedIn) {
+        if (!weakThis)
```

Comment:
Why do we need this check?

Reply:
same as above comment. Please let me know if there is anything wrong with it.